pkg/controller.v1alpha3/suggestion/suggestion_controller.go

```diff
@@ -0,0 +1,176 @@
+		IsController: true,
+		OwnerType:    &suggestionsv1alpha3.Suggestion{},
+	})
+	if err != nil {
+		return err
+	}
+
+	err = c.Watch(&source.Kind{Type: &corev1.Service{}}, &handler.EnqueueRequestForOwner{
+		IsController: true,
+		OwnerType:    &suggestionsv1alpha3.Suggestion{},
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+var _ reconcile.Reconciler = &ReconcileSuggestion{}
+
+// ReconcileSuggestion reconciles a Suggestion object
+type ReconcileSuggestion struct {
+	client.Client
+	composer.Composer
+	suggestionclient.SuggestionClient
+
+	scheme *runtime.Scheme
+}
+
+// Reconcile reads that state of the cluster for a Suggestion object and makes changes based on the state read
+// and what is in the Suggestion.Spec
+// TODO(user): Modify this Reconcile function to implement your Controller logic.  The scaffolding writes
+// a Deployment as an example
+// Automatically generate RBAC rules to allow the Controller to read and write Deployments
+// +kubebuilder:rbac:groups=apps,resources=deployments,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=apps,resources=deployments/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=katib.kubeflow.org,resources=suggestions,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=katib.kubeflow.org,resources=suggestions/status,verbs=get;update;patch
+func (r *ReconcileSuggestion) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	logger := log.WithValues("Suggestion", request.NamespacedName)
+	// Fetch the Suggestion instance
+	instance := &suggestionsv1alpha3.Suggestion{}
+	err := r.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+	oldS := instance.DeepCopy()
+
+	deploy, err := r.DesiredDeployment(instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	_, err = r.createOrUpdateDeployment(deploy)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	service, err := r.DesiredService(instance)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+	_, err = r.createOrUpdateService(service)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	experiment := &experimentsv1alpha3.Experiment{}
+	trials := &trialsv1alpha3.TrialList{}
+
+	if err := r.Get(context.TODO(), types.NamespacedName{
+		Name:      instance.Name,
+		Namespace: instance.Namespace,
+	}, experiment); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if err := r.List(context.TODO(),
+		client.MatchingLabels(util.TrialLabels(experiment)), trials); err != nil {
+		return reconcile.Result{}, err
+	}
+
+	logger.Info("Sync assignments", "suggestions", instance.Spec.Suggestions)
+	if err = r.SyncAssignments(instance, experiment, trials.Items); err != nil {
```

Comment:
Where is this function defined?

Reply:
https://github.com/kubeflow/katib/pull/742/files#diff-1cdb066fdd3258bfce6a7218681d43b6R14

Now we only have a fake implementation. The real implementation should be added after GRPC API is updated. Ref https://github.com/kubeflow/katib/pull/735/files#diff-5ad38cda167bce34ad354bdef2f05d51R36